File: app/src/main/java/com/mobile/model/SeatInfo.java
package com.mobile.model;

/**
 * Created by anubis on 6/27/17.
 */

import org.parceler.Parcel;

import java.util.Random;

@Parcel
public class SeatInfo implements Comparable<SeatInfo> {

    public enum SeatType {
        SeatTypeUnknown,
        SeatTypeCanReserve,
        SeatTypeCanReserveLeft,
        SeatTypeCanReserveRight,
        SeatTypeWheelchair,
        SeatTypeCompanion,
        SeatTypeNotASeat,
        SeatTypeSofaLeft,
        SeatTypeSofaMiddle,
        SeatTypeSofaRight


    }

    boolean available;
    int row;
    int column;
    String seatName;
    String type;

    public SeatInfo() {
    }

    public boolean isAvailable() {
        return available;
    }

    public void setAvailable(boolean available) {
        this.available = available;
    }

    public int getRow() {
        return row;
    }

    public void setRow(int row) {
        this.row = row;
    }

    public int getColumn() {
        return column;
    }

    public void setColumn(int column) {
        this.column = column;
    }

    public String getSeatName() {
        return seatName;
    }

    public void setSeatName(String seatName) {
        this.seatName = seatName;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public SeatType getSeatType() {

        SeatType seatType = SeatType.SeatTypeUnknown;

        if (type.toLowerCase().matches("canreserve")) {
            seatType = SeatType.SeatTypeCanReserve;
        }
        if (type.toLowerCase().matches("canreserveleft")) {
            seatType = SeatType.SeatTypeCanReserveLeft;
        }
        if (type.toLowerCase().matches("canreserveright")) {
            seatType = SeatType.SeatTypeCanReserveRight;
        } else if (type.toLowerCase().matches("wheelchair")) {
            seatType = SeatType.SeatTypeWheelchair;
        } else if (type.toLowerCase().matches("companion")) {
            seatType = SeatType.SeatTypeCompanion;
        } else if (type.toLowerCase().matches("notaseat")) {
            seatType = SeatType.SeatTypeNotASeat;
        } else if (type.toLowerCase().matches("sofaleft")) {
            seatType = SeatType.SeatTypeSofaLeft;
        } else if (type.toLowerCase().matches("sofaright")) {
            seatType = SeatType.SeatTypeSofaRight;
        } else if (type.toLowerCase().matches("sofamiddle")) {
            seatType = SeatType.SeatTypeSofaMiddle;
        }
        return seatType;
    }

    public boolean isWheelChairOrCompanion() {
        SeatType type = getSeatType();
        return type == SeatType.SeatTypeWheelchair || type == SeatType.SeatTypeCompanion;
    }

    @Override
    public int compareTo(SeatInfo another) {
        SeatInfo ss = (SeatInfo) another;

        if (row == ss.row)
            return this.column - ss.column;
        else
            return this.row - ss.row;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        SeatInfo seatInfo = (SeatInfo) o;

        if (row != seatInfo.row) return false;
        if (column != seatInfo.column) return false;
        if (seatName != null ? !seatName.equals(seatInfo.seatName) : seatInfo.seatName != null)
            return false;
        return type != null ? type.equals(seatInfo.type) : seatInfo.type == null;
    }

    @Override
    public int hashCode() {
        int result = row;
        result = 31 * result + column;
        result = 31 * result + (seatName != null ? seatName.hashCode() : 0);
        result = 31 * result + (type != null ? type.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "SeatInfo{" +
                "available=" + available +
                ", row=" + row +
                ", column=" + column +
                ", seatName='" + seatName + '\'' +
                ", type='" + type + '\'' +
                '}';
    }
}

File: app/src/main/java/com/mobile/model/SeatInfo.kt
package com.mobile.model

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.android.parcel.Parcelize

@Parcelize
class SeatInfo(@SerializedName("available") var isAvailable: Boolean = false,
               var row: Int = 0,
               var column: Int = 0,
               var seatName: String? = null,
               var type: String? = null
) : Parcelable, Comparable<SeatInfo> {

    val seatType: SeatType
        get() {

            var seatType = SeatType.SeatTypeUnknown

            if (type!!.toLowerCase().matches("canreserve".toRegex())) {
                seatType = SeatType.SeatTypeCanReserve
            }
            if (type!!.toLowerCase().matches("canreserveleft".toRegex())) {
                seatType = SeatType.SeatTypeCanReserveLeft
            }
            if (type!!.toLowerCase().matches("canreserveright".toRegex())) {
                seatType = SeatType.SeatTypeCanReserveRight
            } else if (type!!.toLowerCase().matches("wheelchair".toRegex())) {
                seatType = SeatType.SeatTypeWheelchair
            } else if (type!!.toLowerCase().matches("companion".toRegex())) {
                seatType = SeatType.SeatTypeCompanion
            } else if (type!!.toLowerCase().matches("notaseat".toRegex())) {
                seatType = SeatType.SeatTypeNotASeat
            } else if (type!!.toLowerCase().matches("sofaleft".toRegex())) {
                seatType = SeatType.SeatTypeSofaLeft
            } else if (type!!.toLowerCase().matches("sofaright".toRegex())) {
                seatType = SeatType.SeatTypeSofaRight
            } else if (type!!.toLowerCase().matches("sofamiddle".toRegex())) {
                seatType = SeatType.SeatTypeSofaMiddle
            }
            return seatType
        }

    val isWheelChairOrCompanion: Boolean
        get() {
            val type = seatType
            return type == SeatType.SeatTypeWheelchair || type == SeatType.SeatTypeCompanion
        }

    enum class SeatType {
        SeatTypeUnknown,
        SeatTypeCanReserve,
        SeatTypeCanReserveLeft,
        SeatTypeCanReserveRight,
        SeatTypeWheelchair,
        SeatTypeCompanion,
        SeatTypeNotASeat,
        SeatTypeSofaLeft,
        SeatTypeSofaMiddle,
        SeatTypeSofaRight
    }

    override fun compareTo(another: SeatInfo): Int {
        val ss = another

        return if (row == ss.row)
            this.column - ss.column
        else
            this.row - ss.row
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false

        val seatInfo = other as SeatInfo?

        if (row != seatInfo!!.row) return false
        if (column != seatInfo.column) return false
        if (if (seatName != null) seatName != seatInfo.seatName else seatInfo.seatName != null)
            return false
        return if (type != null) type == seatInfo.type else seatInfo.type == null
    }

    override fun hashCode(): Int {
        var result = row
        result = 31 * result + column
        result = 31 * result + if (seatName != null) seatName!!.hashCode() else 0
        result = 31 * result + if (type != null) type!!.hashCode() else 0
        return result
    }

    override fun toString(): String {
        return "SeatInfo{" +
                "available=" + isAvailable +
                ", row=" + row +
                ", column=" + column +
                ", seatName='" + seatName + '\''.toString() +
                ", type='" + type + '\''.toString() +
                '}'.toString()
    }

    fun asPosition(): SeatPosition {
        return SeatPosition(row = row, column = column)
    }
}